extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/input/table/SegmentWithMetadata.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.msq.input.table;
+
+import org.apache.druid.segment.Segment;
+import org.apache.druid.timeline.DataSegment;
+
+import javax.annotation.Nullable;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Contains the {@link DataSegment} and {@link Segment}. The datasegment could be null if the segment is a dummy, such
+ * as those created by {@link org.apache.druid.msq.input.inline.InlineInputSliceReader}.
+ */
+public class SegmentWithMetadata implements Closeable
```

Comment:
@findingrish Do you think there is an existing class which has these enteries ?

Reply:
There isn't a class which has these entries. 

On the naming aspect, I find classes with similar naming pattern, for example, `SegmentWithDescriptor` encapsulates Segment and RichSegmentDescriptor. 

Classes with similar naming pattern, 
SegmentWithState
SegmentWithDescriptor
DataSegmentWithMetadata
DataSegmentWithLocation 
DataSegmentsWithSchemas 

A segment is represented using two classes, `Segment` which is the body and `DataSegment` which is the metadata, since this class combines the two, how about calling it `CompleteSegment` or `FullSegment`?

Also, I feel this POJO can be kept in the `processing` module for reuse.

Reply:
there is a lot of duplicate information between `Segment` and `DataSegment`, I think the only stuff `DataSegment` has that isn't available in `Segment` is the shardspec and compaction state, and also there is a `Metadata` class associated with `Segment` which makes this name a bit confusing.

Why not just extract what you need from `DataSegment` instead of including the whole thing?

Reply:
Most usecases would require the`DataSegment` object itself to be passed, for example, returning this from a frameProcessor. We could cut down to the few objects which are not duplicated, but we would likely need to recreate the `DataSegment` again somewhere else if we do that.

Reply:
@clintropolis Does `Segment` provide `LoadSpec`, dimensions/metric names, binary version and size somewhere? 

I guess we could convert the segment to a `QueryableIndex` to get the dimensions and metric names, is that what you had in mind?

Reply:
`Segment` doesn't have `LoadSpec` or size or binary version; `LoadSpec` from `DataSegment` is how the `Segment` is made in the first place more or less, which I think is why this composite type seems a bit strange. `DataSegment` is the stuff to tell something to load the actual segment, so it seems kind of funny to drag it into processing. Maybe I don't understand the use case behind these changes enough since `DataSegment` isn't really used at all today for processing beyond loading the data, and it isn't really obvious what the need for `DataSegment` or its stuff like `LoadSpec` and binary version and stuff is to process data once you have the `Segment`.

Reply:
Renamed the class